public/app/features/plugins/ds_edit_ctrl.ts

```diff
@@ -132,7 +132,7 @@ export class DataSourceEditCtrl {
       this.backendSrv
         .withNoBackendCache(() => {
           return datasource
-            .testDatasource()
+            .testDatasource(this)
```

Comment:
Think this isn't needed. Look at [graphite datasource](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/graphite/config_ctrl.ts#L11) for example which are auto detecting graphite versions.

Reply:
@marefr the problem with the graphite approach is that the code will never trigger for new datasources cause new datasources will not have an id yet, so you need to create datasource, save and then visit datasource page again for the code to trigger once, some kind of postsave hook would be nice here.

Reply:
Looking at the [code](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/ds_edit_ctrl.ts#L172), when saving a datasource for the first time the testDatasource function isn't executed.

So your solution should have same problem as mine/graphite. 

I wonder if the Save & Test button should be Save for new datasource and Save & Test for existing datasources. It doesn't make that big a difference, but at least it's clearer that there are no test executed on first save?

Don't have any good solution for these use cases. It's the chicken or egg first kind of problem.

It would be awesome to be able to guide the user, but not possible to do this in a good way right now. I would recommend an enabled checkbox (true/false) for timescaledb instead of tristate. When then testing the datasource you can verify that the timescaledb is properly enabled if the datasource has timescaledb enabled, if not throw error which will be displayed to the user.

Thoughts?

Reply:
@marefr  Actually it is called after saving just not directly in the path you linked but delayed [Code](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/ds_edit_ctrl.ts#L82)
So the button description is correct and tests are executed for new datasources aswell.

Reply:
My bad. You're right. However I still don't like datasource controller having responsibility to update datasource settings. 

I still like my suggestion for now since that's how many other datasources are implemented, elasticsearch index pattern for example which is validated in testDatasource. 

If you want to make a try using postsave hook, be my guest. This is my rough idea for it:
* if a datasource has a onAfterSave function or similar defined, call that after datasource have been saved, but before testDatasource
* provide the current data in function so you can make changes
* return an object to tell that a new save is needed or not to ds_edit_ctrl
* ds_edit_ctrl updates current data and saves the datasource one more time and then calls testDatasource

Reply:
I agree the implementation should probably be changed to make it more clean, but I didnt want to do refactoring to keep the PR small.
Do you want the onAfterSave hook as part of this PR or want me to split it?

Reply:
I would prefer the hook as separate PR